packages/@romejs/js-compiler/lint/rules/emptyMatches.ts

```diff
@@ -5,26 +5,36 @@
  * LICENSE file in the root directory of this source tree.
  */
 
-import {AnyRegExpBodyItem} from '@romejs/js-ast';
+import {AnyRegExpExpression, AnyRegExpBodyItem} from '@romejs/js-ast';
 import {Path, TransformExitResult} from '@romejs/js-compiler';
 import {descriptions} from '@romejs/diagnostics';
 
 function isQuantifiedMinZero(el: AnyRegExpBodyItem): boolean {
   return el.type === 'RegExpQuantified' && el.min === 0;
 }
 
+function lintEmptyMatches(expr: AnyRegExpExpression): boolean {
+  if (expr.type === 'RegExpSubExpression') {
+    return expr.body
+      .map((item) => {
+        if (item.type === 'RegExpGroupNonCapture' || item.type === 'RegExpGroupCapture') {
+          return lintEmptyMatches(item.expression);
+        } else {
+          return isQuantifiedMinZero(item);
+        }
+      })
+      .every(el => el === true);
```

Comment:
There's a lot of intermediate allocations here. Is there a reason that this can't be:

```suggestion
    for (const item of expr.body) {
      let matches = false;
      if (item.type === 'RegExpGroupNonCapture' || item.type === 'RegExpGroupCapture') {
        matches = lintEmptyMatches(item.expression);
      } else {
        matches = isQuantifiedMinZero(item);
      }
      if (!matches) {
        return false;
      }
    }
```

This also reduces the amount of checks that actually need to be done.

Reply:
Yes, this makes more sense. I thought a functional style would be more readable, didn't really think about allocation. I'll make the change.